src/clients/HttpTunneler.h

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (C) 1996-2018 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef SQUID_SRC_CLIENTS_HTTP_TUNNELER_H
+#define SQUID_SRC_CLIENTS_HTTP_TUNNELER_H
+
+#include "base/AsyncCbdataCalls.h"
+#include "base/AsyncJob.h"
+#include "clients/forward.h"
+#include "clients/HttpTunnelerAnswer.h"
+#include "CommCalls.h"
+#if USE_DELAY_POOLS
+#include "DelayId.h"
+#endif
+#include "http/forward.h"
+
+
+class HttpReply;
+class ErrorState;
+class AccessLogEntry;
+typedef RefCount<AccessLogEntry> AccessLogEntryPointer;
+
+namespace Http
+{
+
+/// Establishes an HTTP CONNECT tunnel through a forward proxy.
+///
+/// The caller receives a call back with Http::TunnelerAnswer.
+///
+/// The caller must monitor the connection for closure because this job will not
+/// inform the caller about such events.
+///
+/// This job never closes the connection, even on errors. If a 3rd-party closes
+/// the connection, this job simply quits without informing the caller.
+class Tunneler: virtual public AsyncJob
+{
+    CBDATA_CLASS(Tunneler);
+
+public:
+    /// Callback dialer API to allow Tunneler to set the answer.
+    template <class Initiator>
+    class CbDialer: public CallDialer, public Http::TunnelerAnswer
+    {
+    public:
+        // initiator method to receive our answer
+        typedef void (Initiator::*Method)(Http::TunnelerAnswer &);
+
+        CbDialer(Method method, Initiator *initiator): initiator_(initiator), method_(method) {}
+        virtual ~CbDialer() = default;
+
+        /* CallDialer API */
+        bool canDial(AsyncCall &) { return initiator_.valid(); }
+        void dial(AsyncCall &) {((*initiator_).*method_)(*this); }
+        virtual void print(std::ostream &os) const override {
+            os << '(' << static_cast<const Http::TunnelerAnswer&>(*this) << ')';
+        }
+    private:
+        CbcPointer<Initiator> initiator_; ///< object to deliver the answer to
+        Method method_; ///< initiator_ method to call with the answer
+    };
+
+public:
+    explicit Tunneler(AsyncCall::Pointer &aCallback);
+    Tunneler(const Tunneler &) = delete;
+    Tunneler &operator =(const Tunneler &) = delete;
+
+    /* configuration; too many fields to use constructor parameters */
```

Comment:
"too many" is a personal style choice. 7 is not too many for other classes and several of these look like they could be retrieved from the ALE by the constructor.

Reply:
resolved at 16e3fc3f.
The constructor now has 5 arguments.
I did not used ALE to retrieve other parameters. Inside FwdState we are assuming that ALE may be nil. Even if this is not possible in latest squid I believe that it is better to not rely on ALE.